FirebasePerformance/Sources/FPRNanoPbUtils.m

```diff
@@ -0,0 +1,502 @@
+      cpuReadings[cpuReadingsCount].has_system_time_us = true;
+      cpuReadings[cpuReadingsCount].user_time_us = gaugeData.userTime;
+      cpuReadings[cpuReadingsCount].has_user_time_us = true;
+      cpuReadingsCount++;
+    }
+
+    if ([obj isKindOfClass:[FPRMemoryGaugeData class]]) {
+      FPRMemoryGaugeData *gaugeData = (FPRMemoryGaugeData *)obj;
+      memoryReadings[memoryReadingsCount].client_time_us =
+          gaugeData.collectionTime.timeIntervalSince1970 * USEC_PER_SEC;
+      memoryReadings[memoryReadingsCount].has_client_time_us = true;
+      memoryReadings[memoryReadingsCount].used_app_heap_memory_kb =
+          (int32_t)BYTES_TO_KB(gaugeData.heapUsed);
+      memoryReadings[memoryReadingsCount].has_used_app_heap_memory_kb = true;
+      memoryReadings[memoryReadingsCount].free_app_heap_memory_kb =
+          (int32_t)BYTES_TO_KB(gaugeData.heapAvailable);
+      memoryReadings[memoryReadingsCount].has_free_app_heap_memory_kb = true;
+      memoryReadingsCount++;
+    }
+  }];
+  cpuReadings = realloc(cpuReadings, cpuReadingsCount * sizeof(firebase_perf_v1_CpuMetricReading));
+  memoryReadings =
+      realloc(memoryReadings, memoryReadingsCount * sizeof(firebase_perf_v1_IosMemoryReading));
+
+  gaugeMetric.cpu_metric_readings = cpuReadings;
+  gaugeMetric.cpu_metric_readings_count = cpuReadingsCount;
+  gaugeMetric.ios_memory_readings = memoryReadings;
+  gaugeMetric.ios_memory_readings_count = memoryReadingsCount;
+  return gaugeMetric;
+}
+
+firebase_perf_v1_ApplicationProcessState ApplicationProcessState(FPRTraceState state) {
+  firebase_perf_v1_ApplicationProcessState processState =
+      firebase_perf_v1_ApplicationProcessState_APPLICATION_PROCESS_STATE_UNKNOWN;
+  switch (state) {
+    case FPRTraceStateForegroundOnly:
+      processState = firebase_perf_v1_ApplicationProcessState_FOREGROUND;
+      break;
+
+    case FPRTraceStateBackgroundOnly:
+      processState = firebase_perf_v1_ApplicationProcessState_BACKGROUND;
+      break;
+
+    case FPRTraceStateBackgroundAndForeground:
+      processState = firebase_perf_v1_ApplicationProcessState_FOREGROUND_BACKGROUND;
+      break;
+
+    default:
+      break;
+  }
+
+  return processState;
+}
+
+#ifdef TARGET_HAS_MOBILE_CONNECTIVITY
+CTTelephonyNetworkInfo *NetworkInfo() {
+  static CTTelephonyNetworkInfo *networkInfo;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    networkInfo = [[CTTelephonyNetworkInfo alloc] init];
+  });
+  return networkInfo;
+}
+#endif
+
+#pragma mark - Nanopb creation utilities
+
+/** Converts the network method string to a value defined in the enum
+ *  firebase_perf_v1_NetworkRequestMetric_HttpMethod.
+ *  @return Enum value of the method string. If there is no mapping value defined for the method
+ * firebase_perf_v1_NetworkRequestMetric_HttpMethod_HTTP_METHOD_UNKNOWN is returned.
+ */
+static firebase_perf_v1_NetworkRequestMetric_HttpMethod FPRHTTPMethodForString(
```

Comment:
I would recommend moving the static functions to top to make it easier to reference. The older code suffered on this. Might as well fix it here.

Reply:
Sg! done.